TrafficCapture/trafficReplayer/src/main/java/org/opensearch/migrations/replay/datahandlers/http/HttpJsonTransformer.java

```diff
@@ -4,57 +4,107 @@
+        pipelineOrchestrator.addInitialHandlers(channel.pipeline(), transformer);
     }
 
-    private NettySendByteBufsToPacketHandlerHandler getEndOfConsumptionHandler() {
-        var last = channel.pipeline().last();
-        return (NettySendByteBufsToPacketHandlerHandler) last;
+    private NettySendByteBufsToPacketHandlerHandler getOffloadingHandler() {
+        return Optional.ofNullable(channel).map(c -> (NettySendByteBufsToPacketHandlerHandler) c.pipeline()
+                        .get(RequestPipelineOrchestrator.OFFLOADING_HANDLER_NAME))
+                .orElse(null);
     }
 
     public CompletableFuture<Void> consumeBytes(ByteBuf nextRequestPacket) {
-        chunkSizes.get(chunkSizes.size()-1).add(nextRequestPacket.readableBytes());
+        chunks.add(nextRequestPacket.duplicate().readerIndex(0).retain());
+        chunkSizes.get(chunkSizes.size() - 1).add(nextRequestPacket.readableBytes());
         if (log.isDebugEnabled()) {
             byte[] copy = new byte[nextRequestPacket.readableBytes()];
             nextRequestPacket.duplicate().readBytes(copy);
             log.debug("Writing into embedded channel: " + new String(copy, StandardCharsets.UTF_8));
         }
-        return CompletableFuture.completedFuture(null).thenAccept(x->
-            channel.writeInbound(nextRequestPacket));
+        return CompletableFuture.completedFuture(null).thenAccept(x ->
+                channel.writeInbound(nextRequestPacket));
     }
 
-    @Override
     public CompletableFuture<AggregatedRawResponse> finalizeRequest() {
-        var consumerHandler = getEndOfConsumptionHandler();
+        var offloadingHandler = getOffloadingHandler();
         channel.close();
-        if (handlerStatus != HTTP_CONSUMPTION_STATUS.DONE) {
-            return CompletableFuture.failedFuture(new MalformedRequestException());
+        if (offloadingHandler == null) {
+            // the NettyDecodedHttpRequestHandler gave up and didn't bother installing the baseline handlers -
+            // redrive the chunks
+            return redriveWithoutTransformation(pipelineOrchestrator.packetReceiver);
         }
-        return consumerHandler.packetReceiverCompletionFuture;
+        return offloadingHandler.getPacketReceiverCompletionFuture()
+                .handle((v, t) -> {
+                    if (t != null) {
+                        if (t instanceof NoContentException) {
+                            return redriveWithoutTransformation(offloadingHandler.packetReceiver);
+                        } else {
+                            throw new CompletionException(t);
+                        }
+                    } else {
+                        return CompletableFuture.completedFuture(v);
+                    }
+                }).thenCompose(Function.identity());
+    }
+
+    private CompletableFuture<AggregatedRawResponse> redriveWithoutTransformation(IPacketToHttpHandler packetConsumer) {
```

Comment:
Is this just simply sending the bytebufs directly to NettyPacketToHttpHandler?

Reply:
Yes, provided that the transformedPacketReceiver was a NettyPacketToHttpHandler.   It also makes sure that it does them so that packets won't be committed out of order.